kernel/src/syscall.rs

```diff
@@ -19,3 +21,65 @@ crate enum Syscall {
     /// Various memory operations.
     MEMOP = 4,
 }
+
+/// Why the process stopped executing and execution returned to the kernel.
+pub enum ContextSwitchReason {
+    /// Process exceeded its timeslice, otherwise catch-all.
+    Other,
+    /// Process called a syscall.
+    SyscallFired,
+    /// Process triggered the hardfault handler.
+    Fault,
+}
+
+/// This trait must be implemented by the architecture of the chip Tock is
+/// running on. It allows the kernel to manage processes in an
+/// architecture-agnostic manner.
+pub trait SyscallInterface {
+    /// Some architecture-specific struct containing per-process state that must
+    /// be kept while the process is not running.
+    type StoredState: Default;
+
+    /// Allows the kernel to query to see why the process stopped running. This
+    /// function can only be called once to get the last state of the process
+    /// and why the process context switched back to the kernel.
+    ///
+    /// An implementor of this function is free to reset any state that was
+    /// needed to gather this information when this function is called.
+    fn get_context_switch_reason(&self) -> ContextSwitchReason;
+
+    /// Get the syscall that the process called.
+    fn get_syscall_number(&self, stack_pointer: *const usize) -> Option<Syscall>;
```

Comment:
These methods should probably all be marked `unsafe`. Fundamentally the expose unsafe interfaces. For example, this one allows the caller to read arbitrary memory.

Reply:
What about extending the `Syscall` enum to include arguments in each of the variants (see below) and combining this method with `get_syscall_data`:

```rust
pub enum Syscall {
     YIELD,
     SUBSCRIBE((usize, usize, usize, usize)), // these could potentially be even more typed, but for another time
     COMMAND((usize, usize, usize, usize)),
     ALLOW((usize, usize, usize, usize)),
     MEMOP((usize, usize, usize, usize)),
}

fn get_syscall(&self, stack_pointer: *const usize) -> Option<Syscall>;
```

Reply:
Do you have a preference of the `(` version versus the `{` version? (I don't know what rust calls the two versions of enum args.)